conformance/tests/httproute.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/roundtripper"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+var ConformanceTests []suite.ConformanceTest = []suite.ConformanceTest{{
+	ShortName:   "HTTPRouteSimpleSameNamespace",
+	Description: "A single HTTPRoute in the gateway-conformance-infra namespace",
+	Manifests:   []string{"httproute-simple-same-namespace.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := v1alpha2.Namespace("gateway-conformance-infra")
+		routeNN := types.NamespacedName{Name: "gateway-conformance-infra-test", Namespace: string(ns)}
+		gwNN := types.NamespacedName{Name: "gateway-same-namespace", Namespace: string(ns)}
+		gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeReady(t, suite.Client, suite.ControllerName, gwNN, routeNN)
+
+		t.Run("Simple HTTP request should reach infra-backend", func(t *testing.T) {
+			t.Logf("Making request to http://%s", gwAddr)
```

Comment:
Good starting point. We may need to have a CLI arg for gwAddr. Just leaving a note, no action is necessary yet.

Reply:
I'd been working with the assumption that Gateway status addresses would always be set on a ready Gateway. Would that not be possible for some of the implementations relying on provisioning Gateways in advance of running these tests?

Reply:
My recommendation from an organization perspective is to try to avoid nesting by resource type in the test cases, as we will end up having conformance things that are split between gwc, gw, and *route behaviors.

instead put that in the name of the test go file. 
Also, if we could have the tests be encapsulated into a file each makes it easier to understand organizationally:

HTTPRouteSimpleSameNamespace_test.go:

```
func init() {
  ConformanceTests = append(ConformanceTests, &ConformanceTest{ ...
  })
}
```

Reply:
@bowei thanks for the suggestion! I've reworked this to use that structure and I think it will make it much easier to add future test cases + understand existing ones.